### PR TITLE
Use code-inspectors to prevent optimization in unsafe contexts.

### PR DESCRIPTION
**Pull request by thinkmoore**
The optimizer should only run when the code being compiled could
directly access racket/unsafe/ops. This prevents unsoundness in Typed
Racket from giving untrusted code access to dangerous operations.

### CODE REVIEW

File: typed-racket-lib/typed-racket/tc-setup.rkt [hunk 1]
```diff
@@ -17,9 +17,18 @@
...
+(define (authorized-code-inspector?)
+  ;; Check whether the current code inspector has sufficient
+  ;; privileges to access unsafe operations.
+  (with-handlers ([exn:fail? (lambda (e) #f)])
+    (module->namespace 'racket/unsafe/ops)
+    #t))
+
 (define (maybe-optimize body)
   ;; do we optimize?
-  (if (and (optimize?) (not (getenv "PLT_TR_NO_OPTIMIZE")))
```
Comment (samth) [inline typed-racket-lib/typed-racket/tc-setup.rkt]: I think we should keep the environment variable -- it might be useful for other things.

Comment (thinkmoore) [inline typed-racket-lib/typed-racket/tc-setup.rkt]: Added this back in and made note of it in the documentation.

**Comment by mflatt** [comment]
I'm not sure, but I think
```
(inspector-superior?
 (current-code-inspector)
 (make-inspector
  (variable-reference->module-declaration-inspector
   (#%variable-reference))))
```
might be a more appropriate test. That should check whether the current code inspector allows access to the enclosing TR module, which I think should be the same as whether the current code inspector allows access to unsafe operations. (Otherwise, references generated by the TR optimizer won't work.)

**Comment by thinkmoore** [comment]
I originally had written (a hackier) version of that check. The reason I switched to trying to access racket/unsafe/ops was that it has the side effect that if for some reason typed racket was loaded in a context that didn't have access to racket/unsafe/ops, it meant that typed racket would silently turn off optimization and still work, which seemed nice.

**Comment by mflatt** [comment]
Ok. I don't see any problem with that strategy.

**Pull request merged**